### DISCUSSION TITLE
TP 10 - Consulta sobre complejidad.

### DISCUSSION THREAD

**Discussion by LautaroTummino** [INGCOM-UNRN-P1/cursada-2023#211] [Preguntas y respuestas] [answered]
```C
int suma_lenta(int numero, int otro_numero)
{
    if (otro_numero > 0) // O(1)
    {
        while (otro_numero) // O(n) -> O(otro_numero) 
        {
            otro_numero = otro_numero - 1; //O(n)
            numero = numero + 1; //O(n)
        }
    }
    else if (otro_numero < 0) // 0(1)
    {
        while (otro_numero) // O(n) -> O(otro_numero)
        {
            otro_numero = otro_numero + 1; //O(n)
            numero = numero - 1; //O(n)
        }
    }
    return numero; //O(1)
}
```

# Función `suma_lenta`

Esta función se encarga de realizar una suma lenta entre dos números ingresados como argumentos

## Complejidad Temporal
La complejidad temporal es de O(n) donde n hasta el 'otro_numero' ??

El costo computacional es de O(1) + O(n) + O(1) + O(n) + O(1). -> Si contabilizo todas las lineas?
                              if -   while - else if -  while -  return
El costo computacional es de O(1) + O(n) + O(1). -> Si contabilizo un solo camino
                              if  -   while -   return

## Complejidad Espacial
La función analizada recibe dos `int` como argumentos, y retorna otro `int`
La complejidad espacial esta en `3 int`'s 

## Consultas

Entiendo que a la hora de analizar la complejidad, en caso de un if/else analizamos solo el camino mas costoso.
Mi consulta  es si estoy por buen camino o si estoy pifiando fuerte con el ejemplo de suma_lenta.

while (otro_numero) // O(n) -> O(otro_numero) , en este ejemplo es O(n) ya que tenemos un lazo simple, pero esto depende del tamaño que tiene otro_numero o no es relevante?

**Comment by AnGys09** [comment]
Por otro lado, dentro de este mismo codigo, lo que esta dentro del while
```C
        while (otro_numero)       // O(N) -> O(otro_numero)
        {
            otro_numero = otro_numero + 1;     //O(N)  o es O(1)
            numero = numero - 1;         //O(N)       o es  O(1)
        }
    }
    return numero; //O(1)
```
Va con la misma notacion de O(N) por que pertenece a ese lazo , o lo tomamos de forma individual como O(1), por ser operaciones aritmecas?

**Reply by LautaroTummino** [reply]
Me parece que se toma el O(n) y lo que esta dentro del while se omite, xq se hace 'otro_numero' veces?  y luego si contabiliza el retorno.
Quedaría como: `O(1) + O(n)`

Me parece que, por como hace la suma del costo computacional en el `tp1-ej5.md` que da el profe en la función `cociente_lento`  entiendo que no contabiliza las dos operaciones aritméticas dentro del while.

**Reply by AnGys09** [reply]
Todavia no entiendo complejidad temporal, complejidad espacial, (ambas) con respecto a costo computacional, por todo lo que vi en el ejemplo que nos dieron, me da a entender que son cosas diferentes, pero la verdad que no se.....

**Comment by TeoInMaghine** [comment]
> Todavia no entiendo complejidad temporal, complejidad espacial, (ambas) con respecto a costo computacional, por todo lo que vi en el ejemplo que nos dieron, me da a entender que son cosas diferentes, pero la verdad que no se.....

El costo computacional es más granular, te podría dar $20n + 15$ por ejemplo, en vez de $O(n)$. Contás cuantas asignaciones, operaciones aritméticas, comparaciones etc a grandes rasgos, y lo sumás. En el caso que algo se ejecute dentro de un lazo que se ejecuta en el peor caso $n$ veces, su costo se multiplica por $n$.

> La complejidad temporal es de O(n) donde n hasta el 'otro_numero' ??

En el caso que preguntás, en efecto la complejidad sería $O(n)$, $n$ es el tamaño de nuestra entrada, en este caso `otro_numero`. Lo importante de la complejidad es la forma de la función, y este concepto incluso es parte de la sintaxis. Por ejemplo $O(n + 1)$ se puede simplificar: $O(n + 1) = O(n)$

Otros ejemplos: 
* $O(156874) = O(1)$
* $O(30n) = O(n)$
* $O(15n^3 + 15) = O(n^3)$. 

Usando la notación de complejidad temporal se simplifica, para darte simplemente la forma en que crece el costo computacional basado en el tamaño de la entrada.

**Reply by LautaroTummino** [reply]
Gracias por el aporte compa!

**Comment by martinvilu** [comment, marked as answer]
Ojo, cuando mencionas `O(n)`, ese `n` tiene que ser una variable del código que estás analizando. Sinó, ¿qué valor es? ¿`O(numero)` u `O(otro_numero)`, o incluso O(numero+otro_numero)`? ¿La complejidad es la misma?

**Comment by martinvilu** [comment]
¿Quedo resuelto esto?

**Reply by LautaroTummino** [reply]
Si gracias!
